### PR TITLE
i18n

### LINKED ISSUE #16: i18n
Module is not internationalized.

I can provide italian translations of the fields if needed.

### PR DESCRIPTION
**Pull request by digitall-it**
Didn't test the code, fixes #16.

### CODE REVIEW

**Comment by lerni** [comment]
I've also created a translation (de), but wasn't ready to PR. Unfortunately @digitall-it PR crashes because of static $cms_message is set to the translation-string (https://github.com/digitall-it/silverstripe-share-care/blob/master/code/ShareCare.php#L43). I think "$cms_message" should be removed and "_t"-translation should be used - right?

Do you @jonom prefer uppercase translation keys?

I also think it would make sense to add default lang-strings to the _t-functions, so you can auto-generate en.yml

I would be willing to merge @digitall-it italian and my german translation into one PR.

**Comment by digitall-it** [comment]
That was my only fear, that statics and translation strings wouldn't mix.

> I think `"$cms_message"` should be removed and `"_t"`-translation should be used

How to do that?

**Comment by lerni** [comment]
@digitall-it I would just reference _t('ShareCare.CMSMessage') instead of `$cms_message` where ever it is used.

File: code/ShareCare.php [hunk 1]
```diff
@@ -40,7 +40,7 @@ class ShareCare extends DataExtension
      * @var string
      * @config
      */
-    private static $cms_message = 'When this page is shared by people on social media it will look something like this:';
+    private static $cms_message = _t('ShareCare.CMSMessage');
```
Comment (jonom) [inline code/ShareCare.php]: As @lerni suggested, this line can be deleted, and wherever `$cms_message` was used, use `_t('ShareCare.CMSMessage')` instead.

File: code/ShareCare.php [hunk 2]
```diff
@@ -49,10 +49,10 @@ public function updateCMSFields(FieldList $fields)
     {
         $msg = Config::inst()->get('ShareCare', 'cms_message');
         if ($msg) {
-            $fields->addFieldToTab('Root.Share', new LiteralField('ShareCareMessage',
+            $fields->addFieldToTab(_t('ShareCare.TabName'), new LiteralField('ShareCareMessage',
```
Comment (jonom) [inline code/ShareCare.php]: Please add a default EN string to each `_t()` call, e.g. `_t('ShareCare.TabName', 'Share')` here.

File: code/ShareCare.php [hunk 3]
```diff
@@ -197,8 +197,8 @@ public function EmailShareLink()
             return false;
         }
         $pageURL = $this->owner->AbsoluteLink();
-        $subject = rawurlencode('Thought you might like this');
-        $body = rawurlencode("Thought of you when I found this: $pageURL");
+        $subject = rawurlencode(_t('ShareCare.EmailSubject'));
+        $body = rawurlencode(_t('ShareCare.Body') . " " . $pageURL);
```
Comment (jonom) [inline code/ShareCare.php]: Can you please revise this to pass the $pageURL variable in to the translation function. See third example here: https://docs.silverstripe.org/en/3/developer_guides/i18n/#usage-in-php-files

File: code/ShareCareFields.php [hunk 4]
```diff
@@ -30,24 +30,24 @@ public function updateCMSFields(FieldList $fields)
...
-        $fields->addFieldToTab('Root.Share', TextAreaField::create('OGDescriptionCustom', 'Share description')
+        $fields->addFieldToTab(_t('ShareCare.TabName'), TextAreaField::create('OGDescriptionCustom', _t('ShareCareFields.ShareDescription'))
             ->setAttribute('placeholder', $this->owner->getDefaultOGDescription())
             ->setRows(2));
-        $fields->addFieldToTab('Root.Share', UploadField::create('OGImageCustom', 'Share image')
+        $fields->addFieldToTab(_t('ShareCare.TabName'), UploadField::create('OGImageCustom', _t('ShareCareFields.ShareImage'))
             ->setAllowedFileCategories('image')
             ->setAllowedMaxFileNumber(1)
-            ->setDescription('<a href="https://developers.facebook.com/docs/sharing/best-practices#images" target="_blank">Optimum image ratio</a> is 1.91:1. (1200px wide by 630px tall or better)'));
+            ->setDescription(_t('ShareCareFields.ShareImageRation')));
```
Comment (jonom) [inline code/ShareCareFields.php]: Typo here - can you please remove the trailing `n` so it is `ShareImageRatio`

File: code/ShareCareFields.php [hunk 5]
```diff
@@ -30,24 +30,24 @@ public function updateCMSFields(FieldList $fields)
...
         if (Config::inst()->get('ShareCare', 'pinterest')) {
-            $fields->addFieldToTab('Root.Share', UploadField::create('PinterestImageCustom', 'Pinterest image')
+            $fields->addFieldToTab(_t('ShareCare.TabName'), UploadField::create('PinterestImageCustom', _t('ShareCareFields.SharePinterestImage'))
```
Comment (jonom) [inline code/ShareCareFields.php]: Can we shorten this to `ShareCareFields.PinterestImage` please

File: code/ShareCareFields.php [hunk 6]
```diff
@@ -30,24 +30,24 @@ public function updateCMSFields(FieldList $fields)
...
                 ->setAllowedFileCategories('image')
                 ->setAllowedMaxFileNumber(1)
-                ->setDescription('Square/portrait or taller images look best on Pinterest. This image should be at least 750px wide.'));
+                ->setDescription(_t('ShareCareFields.SharePinterestDescription')));
```
Comment (jonom) [inline code/ShareCareFields.php]: Shorten to `ShareCareFields.PinterestImageDescription` please

File: code/ShareCareSingleSummary.php [hunk 7]
```diff
@@ -23,15 +23,15 @@ public function updateCMSFields(FieldList $fields)
         // Remove all Meta fields
         $fields->removeByName('Metadata');
         // Add summary fields
-        $fields->addFieldToTab('Root.Main', TextAreaField::create('MetaDescription', 'Content summary')
-            ->setDescription('Summarise the content of this page. This will be used for search engine results and social media so make it enticing.')
+        $fields->addFieldToTab('Root.Main', TextAreaField::create('MetaDescription', _t('ShareCareSummary.DescriptionTitle'))
```
Comment (jonom) [inline code/ShareCareSingleSummary.php]: I think I like `ShareCareSummary.SummaryTitle` better

File: code/ShareCareSingleSummary.php [hunk 8]
```diff
@@ -23,15 +23,15 @@ public function updateCMSFields(FieldList $fields)
...
+            ->setDescription(_t('ShareCareSummary.DescriptionDescription'))
```
Comment (jonom) [inline code/ShareCareSingleSummary.php]: So this one can be `ShareCareSummary.SummaryDescription`

File: templates/ShareCarePreview.ss [hunk 9]
```diff
@@ -84,14 +84,14 @@
...
-			<% if $IncludeTwitter %><a href="https://cards-dev.twitter.com/validator" target="_blank">(debug)</a><% end_if %>
+			<% if $IncludeTwitter %><a href="https://cards-dev.twitter.com/validator" target="_blank">(<%t ShareCare.Debug "Debug" %>)</a><% end_if %>
 			<span class="separator">|</span> <a href="$GooglePlusShareLink" target="_blank">Google+</a>
 			<span class="separator">|</span> <a href="$PinterestShareLink" target="_blank">Pinterest</a>
 			<span class="separator">|</span> <a href="$LinkedInShareLink" target="_blank">LinkedIn</a>
 		</p>
 
 	</div>
-	<span class="description">Previews will be updated after saving. Changes need to be published before they take effect.</span>
+	<span class="description"><%t ShareCare.Description "Previews will be updated after saving. Changes need to be published before they take effect." %></span>
```
Comment (jonom) [inline templates/ShareCarePreview.ss]: Can we call this `ShareCare.SaveNotice` please

File: code/ShareCareFields.php [hunk 10]
```diff
@@ -21,7 +21,7 @@ class ShareCareFields extends DataExtension
      * @var string
      * @config
      */
-    private static $cms_message = 'The preview above is automatically generated from your content. You can override the default values using these fields:';
+    private static $cms_message = _t('ShareCareFields.CMSMessage');
```
Comment (jonom) [inline code/ShareCareFields.php]: As with previous, please remove this static and reference the translatable string instead in code.

**Review by jonom — CHANGES_REQUESTED** [review]
@digitall-it sorry for the late feedback, I somehow missed this PR when it was opened. Generally this looks good thanks, just a few things to tweak if you don't mind.

**Review by jonom — APPROVED** [review]

File: code/ShareCare.php [hunk 11]
```diff
@@ -35,24 +35,17 @@ class ShareCare extends DataExtension
...
-     */
-    private static $cms_message = 'When this page is shared by people on social media it will look something like this:';
-
-    /**
      * Add a Social Media tab with a preview of share appearance to the CMS.
      */
     public function updateCMSFields(FieldList $fields)
     {
-        $msg = Config::inst()->get('ShareCare', 'cms_message');
+		$msg = _t('ShareCare.CMSMessage');
```
Comment (jonom) [inline code/ShareCare.php]: 1. Please move the default string 'When this page is shared by people on social media it will look something like this:' from $tab to $msg. Default for $tab is 'Share'.
2. Spaces for indentation please on PHP files. A few tabs have been introduced in ShareCareFields.php and ShareCare.php (just find/replace with 4 spaces).

File: code/ShareCare.php [hunk 12]
```diff
@@ -197,8 +190,8 @@ public function EmailShareLink()
             return false;
         }
         $pageURL = $this->owner->AbsoluteLink();
-        $subject = rawurlencode('Thought you might like this');
-        $body = rawurlencode("Thought of you when I found this: $pageURL");
+        $subject = rawurlencode(_t('ShareCare.EmailSubject','Thought you might like this'));
+        $body = rawurlencode(_t('ShareCare.EmailBody','Thought of you when I found this: {URL}', [URL=>$pageURL]));
```
Comment (jonom) [inline code/ShareCare.php]: This causes an error because URL should be wrapped in quotes. Can we go a step further and use PHP5.3 syntax as that's the minimum requirement for SS3.x. `array('URL' => $pageURL)`

File: code/ShareCareFields.php [hunk 13]
```diff
@@ -16,38 +16,31 @@ class ShareCareFields extends DataExtension
...
-        $fields->addFieldToTab('Root.Share', TextAreaField::create('OGDescriptionCustom', 'Share description')
+        $fields->addFieldToTab($tab, TextAreaField::create('OGDescriptionCustom', _t('ShareCareFields.ShareDescription','Share description'))
             ->setAttribute('placeholder', $this->owner->getDefaultOGDescription())
             ->setRows(2));
-        $fields->addFieldToTab('Root.Share', UploadField::create('OGImageCustom', 'Share image')
+        $fields->addFieldToTab($tab, UploadField::create('OGImageCustom', _t('ShareCareFields.ShareImage','Share image'))
             ->setAllowedFileCategories('image')
             ->setAllowedMaxFileNumber(1)
-            ->setDescription('<a href="https://developers.facebook.com/docs/sharing/best-practices#images" target="_blank">Optimum image ratio</a> is 1.91:1. (1200px wide by 630px tall or better)'));
+			->setDescription(_t('ShareCareFields.ShareImageRatio','{Link}Optimum image ratio</a> is 1.91:1. (1200px wide by 630px tall or better)', ['Link'=>'<a href="https://developers.facebook.com/docs/sharing/best-practices#images" target="_blank">'])));
```
Comment (jonom) [inline code/ShareCareFields.php]: PHP 5.3 array syntax please

**Review by jonom — CHANGES_REQUESTED** [review]
Thanks guys. Couple more little things to fix then I'll merge in.

**Comment by digitall-it** [comment]
Thanks to both for helping me learn how to care to share share-care.

**Comment by jonom** [comment]
Thanks @digitall-it and @lerni for this contribution!

**Comment by jonom** [comment]
Shipped in [2.0](https://github.com/jonom/silverstripe-share-care/releases/tag/v2.0.0)
